### PR TITLE
enh: pandas-like semi join

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below.

Disclaimer: I think I never used a semi join before. From the [polars docs](https://docs.pola.rs/user-guide/transformations/joins/#semi-join) this seems like it should do the trick 😁

### CODE REVIEW

**Comment by FBruzzesi** [comment]
CI failure is unrelated 😭

File: tests/frame/test_common.py [hunk 1]
```diff
@@ -205,6 +205,32 @@ def test_anti_join(
...
+)
+@pytest.mark.parametrize(
+    ("join_key", "filter_expr", "expected"),
+    [
+        (["a"], (nw.col("b") > 5), {"a": [2], "b": [6], "z": [9]}),
+        (["b"], (nw.col("b") < 5), {"a": [1, 3], "b": [4, 4], "z": [7, 8]}),
+        (["a", "b"], (nw.col("b") < 5), {"a": [1, 3], "b": [4, 4], "z": [7, 8]}),
+    ],
+)
+def test_semi_join(
```
Comment (MarcoGorelli) [inline tests/frame/test_common.py:225]: as this is a new test, could we use the `constructor` fixture for this one, and maybe create a new `tests/frame/join_test` file that we can later migrate the join tests to?

**Review by MarcoGorelli — COMMENTED** [review]
Yessssss love semi-joins, amazing stuff @FBruzzesi !

Minor comment - hopefully CI is more robust after https://github.com/narwhals-dev/narwhals/pull/456

**Comment by FBruzzesi** [comment]
@MarcoGorelli we have another random attacker! 

Attack!GIF

**Review by MarcoGorelli — APPROVED** [review]
amazing, feel free to merge if you're happy with it!

we should ignore the
```
FAILED tests/frame/schema_test.py::test_dtypes - DeprecationWarning: Passing a BlockManager to DataFrame is deprecated and will raise in a future version. Use public APIs instead.
```
warning in pyproject.toml

**Pull request merged**